### PR TITLE
fix: Support artifact ref from tmpl in UI. Fixes #7587

### LINKED ISSUE #7587: Artifact cannot be fetched from UI when artifact repository is specified using workflow template
## Summary

If [artifact repository](https://argoproj.github.io/argo-workflows/fields/#artifactrepositoryref) is specified with workflow template then even though artifact is saved in artifact repository, It is not retrievable from the UI.

What happened/what you expected to happen?
When artifact is fetched from UI it looks into the default artifact repository instead of the one configured in workflow template.

What version of Argo Workflows are you running?
v3.2.4

## Diagnostics

I have looked into the code and I think this is because the UI only considers the workflow spec when trying to identify the artifact repository.
https://github.com/argoproj/argo-workflows/blob/a94b846e67382252831d44624c2f4b1708f7a30c/server/artifacts/artifact_server.go#L168

I am not sure what should be the right way to do this. Perhaps, the artifact ref should be saved in the status of artifact and then fetched from there just like it fetches the rest of artifact?
https://github.com/argoproj/argo-workflows/blob/a94b846e67382252831d44624c2f4b1708f7a30c/server/artifacts/artifact_server.go#L162 
---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by alexec**
Signed-off-by: Alex Collins <alex_collins@intuit.com>
Fixes #7587

Don't bother creating a PR until you've done this:

* [ ] Run `make pre-commit -B` to fix codegen, lint, and commit message problems.

Create your PR as a draft.

* Your PR needs to pass the required checks before it can be approved. If the check is not required (e.g. E2E tests) it
  does not need to pass.
* Once required tests have passed, you can make it "Ready for review".
* Say how how you tested your changes. If you changed the UI, attach screenshots.

Tips:

* If changes were requested, and you've made them, then dismiss the review to get it looked at again.
* Add you organization to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md) if you like.
* You can ask for help!

### CODE REVIEW

File: server/artifacts/artifact_server.go [hunk 1]
```diff
@@ -165,11 +165,7 @@ func (a *ArtifactServer) returnArtifact(ctx context.Context, w http.ResponseWrit
 		return fmt.Errorf("artifact not found")
 	}
 
-	ref, err := a.artifactRepositories.Resolve(ctx, wf.Spec.ArtifactRepositoryRef, wf.Namespace)
-	if err != nil {
-		return err
-	}
-	ar, err := a.artifactRepositories.Get(ctx, ref)
+	ar, err := a.artifactRepositories.Get(ctx, wf.Status.ArtifactRepositoryRef)
```
[the remark this answers is not in the document]
Comment (alexec) [reply server/artifacts/artifact_server.go:168]: this only addresses bug

**Comment by alexec** [comment]
@terrytangyuan while you're on a roll - could I ask for approval for this PR?

**Comment by terrytangyuan** [comment]
> @terrytangyuan while you're on a roll - could I ask for approval for this PR?

I'll review shortly. This is a simple change but I need to take a closer look at the relevant code.

**Review by terrytangyuan — APPROVED** [review]

File: server/artifacts/artifact_server.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (hamza3202) [reply server/artifacts/artifact_server.go:168]: But then another bug still exists where if artifact location is specified at output level then that artifact is not accessible from UI?

**Pull request merged**